File: src/mk3/PythonPackage/src/tamaf/fiparequestdispatcher.py
from typing import TYPE_CHECKING
from .cyclicbehavior import CyclicBehavior
from .aclmessagetemplate import ACLMessageTemplate
from .performative import Performative
from .fiparequestresponderfordispatcher import FIPARequestResponderForDispatcher

if TYPE_CHECKING:
    from .agent import Agent
    from .aclmessage import ACLMessage

class FIPARequestDispatcher(CyclicBehavior):
    def __init__(self, agent: 'Agent', template: ACLMessageTemplate = None):
        super().__init__(agent)
        self.template = template if template else ACLMessageTemplate(performatives={Performative.REQUEST})

    def Action(self):
        incomingRequest = self.Receive(self.template)
        if incomingRequest:
            responder = self.createResponder(incomingRequest)
            if responder is not None:
                self.agent.AddBehavior(responder)

    def createResponder(self, requestMessage: 'ACLMessage') -> FIPARequestResponderForDispatcher:
        return FIPARequestResponderForDispatcher(self.agent, requestMessage)

File: src/mk1/JavaPackage/src/main/java/tamaf/FIPARequestDispatcher.java
package tamaf;

import java.util.Collections;

public class FIPARequestDispatcher extends CyclicBehavior {
    protected ACLMessageTemplate template;

    public FIPARequestDispatcher(Agent agent) {
        this(agent, null);
    }

    public FIPARequestDispatcher(Agent agent, ACLMessageTemplate template) {
        super(agent);
        if (template != null) {
            this.template = template;
        } else {
            this.template = new ACLMessageTemplate();
            this.template.setPerformatives(Collections.singleton(Performative.REQUEST));
        }
    }

    @Override
    public void action() {
        ACLMessage incomingRequest = receive(template);
        FIPARequestResponderForDispatcher responder = createResponder(incomingRequest);
        if (responder != null) {
            agent.addBehavior(responder);
        }
    }

    public FIPARequestResponderForDispatcher createResponder(ACLMessage request) {
        return new FIPARequestResponderForDispatcher(agent, request);
    }
}
